### ISSUE TITLE
onchange="this.form.submit();" is WCAG violation

### ISSUE THREAD

**Issue by dbltoe** [zencart/zencart#4879]
1.5.8 fresh install 06-14-2022 Zip Download with no mods

The store uses this command 6 times to facilitate the user selecting an item.  At least four of the six violate [WCAG 3.2.2](https://www.w3.org/TR/UNDERSTANDING-WCAG20/consistent-behavior-unpredictable-change.html) as the user is unable to use the keyboard to navigate between selections.

From the home page on a desktop or landscape-viewed tablet (the sideboxes do not appear in mobile regardless of orientation), if you click on Please Select for the Music Genres and then hit the down arrow on your keyboard, you will immediately be taken to the Jazz page. (index.php?main_page=product_music_info&products_id=169).  There is no way to get to the Rock page using a keyboard.

The [Alert: JavaScript Event Found page from ANDI](https://myzencarthost.com/accessibility/andi/help/alerts.html?javascript_event_caution) explains why this is a concern with the first bullet on that page.

An example of proper usage is seen on the All Products page where the user CAN select the method used to sort the products.  Also, it appears that includes/classes/Filters/baseFilter.php is not a problem but, I am not sure.

The affected files are all in the includes/templates/template_default/sideboxes directory and are:
tpl_currencies.php - line 14
tpl_manufacturers_select.php - line 15
tpl_music_genres_select.php - line 15
tpl_record_company_select.php - line 15

In looking over the code, the difference between the All Products dropdown and the Sidebox dropdowns is that the All Products selections are preset and pre-defined whereas a storeowner could have 50 manufacturers.  Also, the dropdown for the sideboxes is based on the ID vs the name of the item.

In some cases, adding a submit button could solve the problem BUT, using only a keyboard one would have to abandon their choice in the dropdown to submit the form with the button.  In the All Products situation, one just picks their choice and hits Enter.  One would have to ensure that the selection would be made with the down arrow and the tab key to take the user directly to the Submit button.  I hate to see the addition of a submit button as it makes selection less "easy" for the mouse users

Unfortunately, my limited coding ability has not been able to make anything work.  Hopefully, one fix can be copied to each file in order to make it work as this is one of the first items that ANDI and WAVE from WebAIM catch.

**Comment by lat9** [comment]
> **_An example of proper usage_** is seen on the All Products page where the user CAN select the method used to sort the products. Also, it appears that includes/classes/Filters/baseFilter.php is not a problem but, I am not sure.

I'm going to disagree here.  On the `all_products` page as well as the product-listing pages, any of the filters that include the `onchange="this.form.submit();"` construct, it's not possible to sort (unless with multiple attempts) by either the first or second filter in the list.  For the products' listing, when both the sort-by and the alpha-filter are in effect, it gets even more "difficult".

**Comment by lat9** [comment]
> Also, the dropdown for the sideboxes is based on the ID vs the name of the item.

Would adding a `title="what is this selecting"` tag to the `select` statements suffice?  Noting that all the underlying processing for those sidebox forms is expecting an ID to be submitted.

**Comment by lat9** [comment]
> Also, it appears that includes/classes/Filters/baseFilter.php is not a problem but, I am not sure.

I'm struggling to find where the classes/Filters files are used.  @zcwilt, please advise.

**Comment by lat9** [comment]
Noting that this issue was previously reported in #4482.  Not sure what we need to do to combine the two varying discussions.

**Comment by dbltoe** [comment]
Perhaps I have incorrectly represented the main problem here.  ADA/WCAG requires that all select box operations with an onChange event be able to be performed by keyboard input only.  

On the All Products Page, you can do all Sort by: selections from the keyboard.

Yet, there is no way you can use the keyboard to get to Rock under Music Genres OR anything other than Fox in Manufacturers.  The user should be able to down-arrow to their selection and press enter on the keyboard.

Perhaps coming up with some way to make the creation of the sidebox dropdowns `$content .= zen_draw_pull_down_menu('music_genre_id', $music_genres_array, (isset($_GET['music_genre_id']) ? $_GET['music_genre_id'] : '')`match the method for the Sort by: `<select name="disp_order" `

I'm just not familiar enough with the process to come up with the answer.

**Comment by lat9** [comment]
@dbltoe, if I traverse a zc157d `products_all` page with the keyboard, using the 'tab' key to wind up on the sort-selector and then attempt to change the sort via the down-arrow key, once I get to "Product Name -desc" the form auto-submits for that selection.

I _do_ understand the issue, I'm just trying to get you to confirm that the issue also exists on those listing-based sort- and filter-dropdowns as well.

**Comment by dbltoe** [comment]
It's a long way to the Sort by: (54 tabs from the jeandret.com start). 😛   It does not make any selection for me on jeandret.com until you hit enter.  Nor does it on my hours-old installation of the latest 1.5.8 at test.feedchaffhaye.com.

The only ones violating the standard are Music Genres, Record Companies, and Manufacturers.  If someone were to clone the construction of these dropdowns (e,g, for recipes), the problem would be inherited in the clone.

To quickly get to Music Genres; on the very latest 1.5.8, go to test.feedchaffhaye.com.com, hit the tab twice, then enter.  Then hit the tab 27 times to get to the Music Genres Please select.  Using the down arrow here DOES auto submit.

**Comment by lat9** [comment]
@dbltoe, that's not what I'm seeing.  If I tab through the various elements until the 'Sort By' select on the "All Products" page and then press the down-arrow key to view the selections, I'm taken directly to the "Products Name - desc" sort.
